python/iot-msk-lambda-pipeline/msk_demo/msk_demo_stack.py

```diff
@@ -0,0 +1,324 @@
+            f"tar -xzf {constants['KAFKA_DOWNLOAD_VERSION']}.tgz",
+            f"./{constants['KAFKA_DOWNLOAD_VERSION']}/bin/kafka-topics.sh --create --zookeeper {zookeeper} --replication-factor 2 --partitions 1 --topic {constants['MSK_TOPIC']}",
+        )
+
+
+class MskBroker(NestedStack):
+    def __init__(self, scope: Construct, construct_id: str, vpc, client_subnet, **kwargs):
+        super().__init__(scope, construct_id, **kwargs)
+
+        # Create the MSK cluster with SASL/SCRAM authentication
+        self.cluster = msk.Cluster(self, "Cluster",
+            cluster_name="iotCluster",
+            kafka_version=msk.KafkaVersion.V2_8_1,
+            vpc=vpc,
+            encryption_in_transit=msk.EncryptionInTransitConfig(
+                client_broker=msk.ClientBrokerEncryption.TLS
+            ),
+            client_authentication=msk.ClientAuthentication.sasl(
+                scram=True
+            ),    
+        )
+
+        # Enable MSK cluster connection  
+        self.cluster.connections.allow_from(
+            ec2.Peer.ipv4("0.0.0.0/0"),
+            ec2.Port.tcp(2181))
+        self.cluster.connections.allow_from(
+            ec2.Peer.ipv4("0.0.0.0/0"),
+            ec2.Port.tcp(9096))
+        
+        # EC2 Instance in the public subnet used to create the topics
+        client = MSKClient(self, "MskClient", 
+            vpc=vpc, 
+            client_subnet=client_subnet, 
+            zookeeper=self.cluster.zookeeper_connection_string)
+
+
+class MskDemoStack(Stack):
+
+    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
+        super().__init__(scope, construct_id, **kwargs)
+
+        # 1 Public Subnet and 2 Private
+        subnets = []
+        subnets.append(ec2.SubnetConfiguration(
+            name = "MSKDemo-subnet-public1", 
+            subnet_type = ec2.SubnetType.PUBLIC, 
+            cidr_mask = 20))
+        subnets.append(ec2.SubnetConfiguration(
+            name = "MSKDemo-subnet-private1", 
+            subnet_type = ec2.SubnetType.PRIVATE_WITH_NAT, 
+            cidr_mask = 20))
+        subnets.append(ec2.SubnetConfiguration(
+            name = "MSKDemo-subnet-private2", 
+            subnet_type = ec2.SubnetType.PRIVATE_ISOLATED, 
+            cidr_mask = 20))
+        
+        vpc = ec2.Vpc(self, "MskVpc",
+            cidr="10.0.0.0/16",
+            nat_gateways=1,
+            max_azs=2,
+            subnet_configuration = subnets)
+
+        # Instantiates the MSK cluster and an EC2 Instance used to create the topics
+        msk_cluster = MskBroker(self, 'MSKBroker', 
+            vpc=vpc, 
+            client_subnet=ec2.SubnetSelection(subnet_type=ec2.SubnetType.PUBLIC)
+        ).cluster
+        
+        # Create MSK cluster secret and encryption key
+        msk_cluster.add_user('demo')
+
+        # Policies that will be applied to Iot Core components and the Lambda function
+        iot_task_policy = iam.PolicyStatement(
```

Comment:
Wow. 

That's a *big* policy that allows a *lot* of stuff.
Could this be done by using the various grant_x(...) calls instead to scope this down to the appropriate resources?